### DISCUSSION TITLE
version 4.26.1 broke my mutations

### DISCUSSION THREAD

**Discussion by gmcamposano** [TanStack/query#5136] [General]
I cannot tell if maybe its a breaking change or a bug as I went through the changelog installing version by version.

```
const saveToDatabase = useMutation(
    async () => {
      await axios.post('/api/cierre', {
        factura: facturaArr,
        datos: {
          ...datos,
          facturado_por: session.user_id,
          month_created: new Date().getMonth(),
          year_created: new Date().getFullYear(),
          tienda: session.tienda
        }
      })
    },
    {
      onSuccess: () => {
        queryClient.invalidateQueries('facturas')
      }
    }
  )
```

I then run in the accept button: `saveToDatabase.mutate()`

checking the req.body from the /api/cierre i get the following:
```
{
  factura: [],
  datos: {
    forma_de_pago: '',
    tipo_de_factura: '',
    nombre_empresa: '',
    rnc_empresa: 0,
    facturado_por: 'clebtvzgc0000hmre7n845d0s',
    month_created: 2,
    year_created: 2023,
    tienda: 'agora'
  }
}
```
What is happening: information arrives incomplete in the api endpoint (factura missing, and ...datos missing too)

on version 4.26.0 I still get all data as expected

**Comment by TkDodo** [comment]
can you please show a codesandbox reproduction?

I'm also wondering why you're not using `variables` as input to the `mutationFn` 🤔

**Reply by gmcamposano** [reply]
I did a minimal example and useMutation kindda works as expected. I would understand if my code was wrong it wouldn't work on other versions, but I went version per version... and 4.26.1 was the one that made useMutation stop updating parts of the data as you see on the example. Was there any breaking changes there that i am not aware off?

**Reply by TkDodo** [reply]
well that version only had one fix, which made sure that options of mutations are kept up to date when components re-render to avoid stale closure issues:

- https://github.com/TanStack/query/pull/5085

not sure how this can lead to the _opposite_ behaviour on your end. maybe your component is re-rendering in between, and variables you closure over, like `facturaArr` (where I don't know where it is coming from because you're not showing that), are being re-set...

**Reply by gmcamposano** [reply]
I am doing some testing and its working using variables as you said before. I am changing the complete function so that when i call createSomething.mutate({...pass data here...})
I seems there is the difference. In versions before 4.26 it allowed me to not use variables, where later on.. i MUST include variables inside mutate func

I can now confirm. It did work now.  Here is how i solved it:
```
  const saveToDatabase = useMutation({
    mutationFn: async (data) => {
      await axios.post('/api/cierre', data)
    },
    onSuccess: () => {
      queryClient.invalidateQueries(['facturas'])
    }
  })
  ```
  ```
  saveToDatabase.mutate({
      factura: facturaArr,
      datos: {
        ...datos,
        facturado_por: session.user_id,
        month_created: new Date().getMonth(),
        year_created: new Date().getFullYear(),
        tienda: session.tienda
      }
    })
    ```

**Reply by TkDodo** [reply]
I would still be very interested in a reproduction of your issue to see if there is something wrong on our end. Theoretically, it should also work without variables...

**Reply by gmcamposano** [reply]
> I would still be very interested in a reproduction of your issue to see if there is something wrong on our end. Theoretically, it should also work without variables...

I will try to make a reproduction of it, as it uses zustand to handle state. I am basically making post requests to the api of states stores with zustand.

**Comment by gmcamposano** [comment]
I will try to create a minimal codesandbox reproduction, but it was only affected from 4.26.1, which works fine in 4.26.0

As to why I am not using variables? Great question. I did not know it was possible. Now i know :)
